nightshift/comms/storage.py

```diff
@@ -0,0 +1,210 @@
+# -*- coding: utf-8 -*-
+
+"""
+This module handles communication with network drive accessible via SFTP where Sierra
+dumps daily files for processing and where CAT staff can access produces MARC files.
+"""
+from io import BytesIO
+import logging
+import os
+from typing import Optional
+
+
+from paramiko.transport import Transport
+from paramiko.sftp_client import SFTPClient
+from paramiko.ssh_exception import SSHException
+
+from ..ns_exceptions import DriveError
+
+
+logger = logging.getLogger("nightshift")
+
+
+def get_credentials():
+    """
+    Retrieves SFTP credentials from environmental variables.
+
+    Returns:
+        credentials
+    """
+    return (
+        os.getenv("SFTP_HOST"),
+        os.getenv("SFTP_USER"),
+        os.getenv("SFTP_PASSW"),
+        os.getenv("SFTP_NS_SRC"),
+        os.getenv("SFTP_NS_DST"),
+        os.getenv("SFTP_PORT"),
+    )
+
+
+class Drive:
+    def __init__(
+        self,
+        host: str,
+        user: str,
+        password: str,
+        src_dir: str,
+        dst_dir: str,
+        port: Optional[str] = None,
+    ) -> None:
+        """
+        Opens a secure communication channel via SFTP to a networked drive
+
+        Args:
+            host:                       SFTP host
+            port:                       SFTP port (optional)
+            user:                       SFTP user name
+            password:                   SFTP user password
+            home_directory:             NighShift directory on the drive
+        """
+        sock = self._sock(host, port)
+        self.sftp = self._sftp(sock, user, password)
+        self.src_dir = src_dir
+        self.dst_dir = dst_dir
+
+    def __enter__(self, *args):
+        return self
+
+    def __exit__(self, *args):
+        self.close()
+
+    def fetch_file(self, src_fh: str) -> Optional[BytesIO]:
+        """
+        Retrieves file of the given path
+
+        Args:
+            path:                       path to file on the SFTP server
+
+        Returns:
+            bytes stream
+
+        Raises:
+            DriveError
+        """
+        src_file_path = self._construct_src_file_path(src_fh)
+        logging.info(f"Fetching {src_file_path} file from the SFTP.")
+        if self.sftp:
+            try:
+                with self.sftp.file(src_file_path, mode="r") as file:
+                    file_size = file.stat().st_size
+                    file.prefetch(file_size)
+                    file.set_pipelined()
```

Comment:
What's the advantage to using `set_pipelined`? I don't stream too many files in my work so I'm curious if there's something that I should be doing.